proposals/0000-cases.md

```diff
@@ -0,0 +1,641 @@
+   can have an arbitrary non-zero number of equations
+ - Lambda expressions cannot have guards
+ - Lambda expressions must have at least one parameter
+
+There have been multiple attempts in the past to bring the capabilities of
+lambda expressions closer to those of function declarations:
+
+ 1. The extension `-XLambdaCase` introduces a `\case` construct which allows
+    lambda expression to have multiple clauses, however, only one pattern can
+    be matched on. Like a regular `case`-expression, this can also have guards.
+    [During its
+    implementation](https://gitlab.haskell.org/ghc/ghc/issues/4359#note_44819)
+    as well as [after it](https://github.com/ghc-proposals/ghc-proposals/pull/18),
+    there were attempts to make it possible to match on multiple patterns. No
+    solution was found, in part because this would make it different from
+    regular case-expressions.
+    - If there were an expression that had pattern matching syntax more similar
+      to lambda expressions but which could also have guards and multiple clauses,
+      it could be used instead of `-XLambdaCase` and would be able to match on
+      multiple patterns.
+ 2. The extension `-XMultiWayIf` essentially introduces standalone guards,
+    simplifying the use of guards that aren't at the outermost level of a
+    function declaration or case-expression. Among other things, this made it
+    easier to use guards inside of lambda expressions.
+    - If there were an expression similar to lambda expressions that could have
+      guards and wasn't required to have at
+      least one parameter, it could be used instead of `-XMultiWayIf`. This
+      includes all uses of `-XMultiWayIf`, not just those inside of lambdas
+      (see Example section).
+ 3. During the implementation of `-XLambdaCase`,
+    [some suggested](https://gitlab.haskell.org/ghc/ghc/issues/4359#note_51110)
+    allowing lambda expressions to have multiple clauses. This was not
+    implemented: The most obvious approach of turning `\` into a layout herald
+    had the disadvantage of making some common idioms invalid.
+    - This could be circumvented by introducing a new expression that isn't required
+      to be backwards compatible with existing idioms.
+
+## Proposed Change Specification
+
+Example:
+
+```haskell
+filter = \cases _ []                 -> []
+                p (x:xs) | p x       -> x : filter p xs
+                         | otherwise ->     filter p xs
+```
+
+A new extension `-XMultiWayLambda` is introduced. Under this extension, a new
+expression is enabled, introduced by the token sequence <tt>\\&nbsp;cases</tt>. The whitespace between  `\` and `cases`
+is optional and may contain an arbitrary sequence of whitespace characters.
+`\cases` behaves in a way largely similar to `\`, but it is a layout herald, can have multiple
+clauses, and may contain guards (see BNF for details).
+
+Zero clauses are not permitted, as the expression would be ambiguous.
```

Comment:
Move the BNF changes up to the top. Then the rest of the section can refer to it.  

Use 'n' for the number of patterns, and 'm' for the number of clauses.  Then you can say

"As the BNF shows
* There can be any number of *patterns* including zero (n >= 0)
* There must be at least one clause (m >= 1).  If there were zero clauses, the arity of the \case would be ambiguous."